### PR TITLE
3750-2

### PR DESCRIPTION
**Pull request by Sufl0wer**
## Номер
3750

## Номер задания
2

## Ссылка на видео с демо
https://youtu.be/QS0uXe8I7CA

## Комментарии
ссылка на видео как запустить бота, описание под видео
https://youtu.be/jJ9l9RVHCgU

### CODE REVIEW

File: 3750/2/checkiner.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  include Telegram::Bot::UpdatesController::MessageContext
+  include StartCommand
+  include CheckinCommand
+  include CheckoutCommand
+  include Helpers
+  include DataCheckConversation
+  include PathGenerator
+  include Saver
+  include GeoValidator
+  include PhotoValidator
```
Comment (kislak) [inline 3750/2/checkiner.rb]: I'd like to see a list of commands availaable:
```
COMMANDS = %i[
  start
  checkin
  checkout
  ...
]
```

File: 3750/2/checkiner.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    super
+    Telegram::Bot::UpdatesController.session_store = :redis_store, { expires_in: 1.month }
+  end
+end
+
+TOKEN = ENV['BOT_TOKEN']
+
+bot = Telegram::Bot::Client.new(TOKEN)
+
+require 'logger'
```
Comment (kislak) [inline 3750/2/checkiner.rb]: requre at top of the file

File: 3750/2/checkiner.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+logger = Logger.new(STDOUT)
+poller = Telegram::Bot::UpdatesPoller.new(bot, WebhooksController, logger: logger)
```
Comment (kislak) [inline 3750/2/checkiner.rb]: define:
bot, controller, logger
and put controller into separate file

File: 3750/2/checkiner.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+require 'json'
+require 'time'
+require 'yaml'
+require_relative 'file_accessor'
+require_relative 'webhooks_controller'
+
+TOKEN = ENV['BOT_TOKEN']
+
+bot = Telegram::Bot::Client.new(TOKEN)
+
```
Comment (kislak) [inline 3750/2/checkiner.rb]: controller = WebhooksController

File: 3750/2/commands/checkin_command.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
+require_relative '../data_check'
```
Comment (kislak) [inline 3750/2/commands/checkin_command.rb]:
```
module Commands
  module Checkin
     def smth
     end
  end
end

include Commands::Checkin
```

File: 3750/2/webhooks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    checkout
+  ].freeze
+
+  def initialize(*)
+    super
+    Telegram::Bot::UpdatesController.session_store = :redis_store, { expires_in: 1.month }
+  end
+
+  def notify(msg)
+    respond_with :message, text: msg
```
Comment (kislak) [inline 3750/2/webhooks_controller.rb]: def notify(:user_not_loggedin)
     respond_with :message, text: MSG[::user_not_loggedin]
end

File: 3750/2/commands/checkin_command.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+module CheckinCommand
+  include DataCheck
```
Comment (kislak) [inline 3750/2/commands/checkin_command.rb]: don't use multiple inheritence (include)
put all helper methods

File: 3750/2/commands/checkin_command.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+  MSG = {
```
Comment (kislak) [inline 3750/2/commands/checkin_command.rb]: put all messages in one place. 
don't split in the classes.

File: 3750/2/path_generator.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
+module PathGenerator
```
Comment (kislak) [inline 3750/2/path_generator.rb]: make it a class

File: 3750/2/data_check.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      proceed_to_geo_check
+      notify(MSG[:photo_check_success])
+    else
+      stay_in_photo_check
+      notify_with_reference(MSG[:photo_check_failure])
+    end
+  end
+
+  def geo_check(*)
+    if geo?
```
Comment (kislak) [inline 3750/2/data_check.rb]: ``` proceed_to_command_ending && return if geo? ```

File: 3750/2/webhooks_controller.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+require_relative 'commands/checkin_command'
+require_relative 'commands/checkout_command'
+
+class WebhooksController < Telegram::Bot::UpdatesController
+  include Telegram::Bot::UpdatesController::MessageContext
+  include StartCommand
+  include CheckinCommand
+  include CheckoutCommand
+
+  COMMANDS = %i[
```
Comment (kislak) [inline 3750/2/webhooks_controller.rb]: delete the code you don't use

File: 3750/2/data_check.rb [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+  def longitude?
+    ALLOWED_LONGITUDE.cover? payload['location']['longitude']
+  end
+
+  def photo?
+    payload['photo']
+  end
+
+  def path
```
Comment (kislak) [inline 3750/2/data_check.rb]: check the naming

File: 3750/2/data_check.rb [hunk 13]
```diff
@@ -0,0 +1,75 @@
+module DataCheck
+  ALLOWED_LATITUDE = 53.914260..53.916240
+  ALLOWED_LONGITUDE = 27.565940..27.571310
+
+  def photo_check(*)
```
Comment (kislak) [inline 3750/2/data_check.rb]: try to split photo_check and geo_check into separate modules

File: 3750/2/commands/checkin.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+    private
+
+    def process_checkin
+      session[:command] = 'checkin'
+      session[:timestamp] = Time.now.getutc.to_i
+      save_context :photo_check
+    end
+
+    def set_checkin_flags
```
Comment (kislak) [inline 3750/2/commands/checkin.rb]: You can probably use one session variable to check if user checkedin:
```
def set_checkin_flags
  session[:checked_in] = true
end
```

File: 3750/2/webhooks_controller.rb [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  def notify_with_reference(msg_key)
+    reply_with :message, text: MSG[msg_key]
+  end
+
+  def send_sticker(msg_key)
+    respond_with :sticker, sticker: MSG[msg_key]
+  end
+
```
Comment (kislak) [inline 3750/2/webhooks_controller.rb]:
``` 
 def registered?
    session.key?(:number)
  end
```

File: 3750/2/webhooks_controller.rb [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  def registered?
+    return true if session.key?(:number)
+    false
+  end
+
+  def checkin?
```
Comment (kislak) [inline 3750/2/webhooks_controller.rb]:
```
  def checked_in?
     session[:checked_in]
  end

  def checked_out?
    !checked_in
  end
```

File: 3750/2/checkiner.rb [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+require 'fileutils'
+require 'open-uri'
+require 'json'
+require 'time'
+require 'yaml'
+require_relative 'file_accessor'
+require_relative 'webhooks_controller'
+require_relative 'path_generator'
+
+MSG = {
```
Comment (kislak) [inline 3750/2/checkiner.rb]: MSG should be defined close to where you use them

File: 3750/2/contexts/photo_check.rb [hunk 18]
```diff
@@ -0,0 +1,38 @@
+module Contexts
+  module PhotoCheck
+    def photo_check(*)
+      proceed_to_geo_check if photo?
+      stay_in_photo_check unless photo?
+    end
+
+    private
+
+    def proceed_to_geo_check
```
Comment (kislak) [inline 3750/2/contexts/photo_check.rb]: can it be?
```
def proceed_to_geo_check
  save_photo
  notify(:photo_check_success)
  save_context(:geo_check)
end
```

Comment (Sufl0wer) [reply 3750/2/contexts/photo_check.rb]: yes, it's possible

File: 3750/2/contexts/photo_check.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
+module Contexts
+  module PhotoCheck
+    def photo_check(*)
+      proceed_to_geo_check if photo?
```
Comment (kislak) [inline 3750/2/contexts/photo_check.rb]: update as in geo_check

File: 3750/2/path_generator.rb [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  def checkin_path
+    "./public/#{id}/checkins/#{timestamp}/"
+  end
+
+  def checkout_path
+    "./public/#{id}/checkouts/#{timestamp}/"
+  end
+
+  def path_request_url
+    BOT_API_URL + GET_PATH_URL + photo_id
```
Comment (kislak) [inline 3750/2/path_generator.rb]: use string interpolation (see 42 line)

File: 3750/2/contexts/photo_check.rb [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+    def proceed_to_geo_check
+      photo_save
+      notify(:photo_check_success)
+      save_context(:geo_check)
+    end
+
+    def photo?
+      payload['photo']
+    end
+
```
Comment (kislak) [inline 3750/2/contexts/photo_check.rb]: create interface like:
```
  PathGenerator.save_photo(session, payload['photo'])
```

move create_path photo_save to PathGenerator

do the same for geo_save
```
  PathGenerator.save_location(session, payload['location'])
```

rename PathGenerator to Storage

Comment (kislak) [reply 3750/2/contexts/photo_check.rb]:
```
storage = Storage.new(session, payload)
storage.save_location

Storage.save_location(session, payload)
```

**Pull request merged**